crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,222 @@
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum SupportedCodeAction {
+    QuickFix,
+    FixAll,
+    #[allow(dead_code)] // TODO: remove
+    OrganizeImports,
+}
+
+#[derive(Clone, Debug)]
+struct DiagnosticEdit {
+    original_diagnostic: types::Diagnostic,
+    diagnostic_fix: DiagnosticFix,
+    document_edits: Vec<types::TextDocumentEdit>,
+}
+
 impl super::BackgroundDocumentRequestHandler for CodeAction {
     super::define_document_url!(params: &types::CodeActionParams);
     fn run_with_snapshot(
         snapshot: DocumentSnapshot,
         _notifier: Notifier,
         params: types::CodeActionParams,
     ) -> Result<Option<types::CodeActionResponse>> {
+        let available_actions = available_code_actions(params.context.only);
+        // fast path - return early if no actions are available
+        if available_actions.is_empty() {
+            return Ok(None);
+        }
         let document = snapshot.document();
-        let url = snapshot.url();
-        let encoding = snapshot.encoding();
-        let version = document.version();
-        let actions: Result<Vec<_>> = params
-            .context
-            .diagnostics
-            .into_iter()
-            .map(|diagnostic| {
-                let Some(data) = diagnostic.data else {
-                    return Ok(None);
-                };
-                let diagnostic_fix: crate::lint::DiagnosticFix = serde_json::from_value(data)
-                    .map_err(|err| anyhow::anyhow!("failed to deserialize diagnostic data: {err}"))
-                    .with_failure_code(lsp_server::ErrorCode::ParseError)?;
-                let edits = diagnostic_fix
-                    .fix
-                    .edits()
-                    .iter()
-                    .map(|edit| types::TextEdit {
-                        range: edit.range().to_range(
-                            document.contents(),
-                            document.index(),
-                            encoding,
-                        ),
-                        new_text: edit.content().unwrap_or_default().to_string(),
-                    });
-
-                let changes = vec![types::TextDocumentEdit {
-                    text_document: types::OptionalVersionedTextDocumentIdentifier::new(
-                        url.clone(),
-                        version,
-                    ),
-                    edits: edits.map(types::OneOf::Left).collect(),
-                }];
-
-                let title = diagnostic_fix
-                    .kind
-                    .suggestion
-                    .unwrap_or(diagnostic_fix.kind.name);
-                Ok(Some(types::CodeAction {
-                    title,
-                    kind: Some(types::CodeActionKind::QUICKFIX),
+        // compute the associated document edits for each diagnostic
+        // these will get re-used when building the actual code actions afterwards
+        let edits = diagnostic_edits(
```

Comment:
I'm struggling a bit with the terminology: 

* A diagnostic has a list of edits
* The diagnostic gets converted into a `DiagnosticEdit` but it itself has a list of edits
* `diagnostic_edits` returns a vec of `DiagnosticEdit` (which is a list of edits). 

My recommendation would have been to rename `DiagnosticEdit` to `DiagnosticFix` but that type already exists and that type also stores a list of edits (but different edits?). 

I don't have a good recommendation on how to simplify this. Maybe rename `DiagnosticFix` to `SerializedFix` or `SerializedDiagnosticFix` to make it clear it's what is stored on `Diagnostic` or `SerializedDiagnosticData`. I mainly want to point out that I'm overwhelmed by this and struggle to understand the different data structures.

Reply:
I've made renamings similar to what you suggested, with one exception: I changed `DiagnosticFix` to `AssociatedDiagnosticData` because I think that describes what its actual role is as a serialized structure. `SerializedFix` may imply that it contains serialized values, which it doesn't.

I also reworked `DiagnosticEdit` (now `DiagnosticFix`) to only include the relevant fields from `AssociatedDiagnosticData` so it's a bit less confusing.